### ISSUE TITLE
Adding new renderers using push_notebook triggers errors

### ISSUE THREAD

**Issue by philippjfr** [bokeh/bokeh#7590]
In earlier versions of bokeh (specifically 0.12.10) it was possible to add new bokeh renderers to a plot and then use ``push_notebook`` to update the plot. This still mostly works fine using recent versions (0.12.13+) but the JS console shows an error during rendering, which appears to be related to the ``ToolbarPanelView`` being deleted:

```javascript
    const toolbar = this._toolbar_views[this.model.toolbar.id];
    toolbar.render();

VM995:11109 Uncaught TypeError: Cannot read property 'render' of undefined  
```

This is because ``build_views`` appears to delete the view just before rendering:

```javascript
 for (const model_id of to_remove) {
  view_storage[model_id].remove()
  delete view_storage[model_id]
}
```

It is worth noting that this works fine on bokeh server so it likely has something to do with some state not being synced in the notebook. I realize that this may not be officially supported but since this used to work it would be very nice if it could be made to work again.

Here is a small example that demonstrates the issue:

```python
from bokeh.plotting import figure, output_notebook
from bokeh.io import  push_notebook

# output to static HTML file
output_notebook()

p = figure(plot_width=400, plot_height=400)

# add a circle renderer with a size, color, and alpha
p.circle([1, 2, 3, 4, 5], [6, 7, 2, 4, 5], size=20, color="navy", alpha=0.5)

handle = show(p, notebook_handle=True)

p.circle([6, 7, 2, 4, 5], [1, 2, 3, 4, 5], size=20, color="red", alpha=0.5)
push_notebook(handle=handle)
```

**Comment by mattpap** [comment]
@philippjfr, does this happen in `master`. I fixed a similar issue recently.

**Comment by philippjfr** [comment]
Yes, it does unless I messed something up. I'd be happy to do a bisect on this but probably won't get to it today.

**Comment by philippjfr** [comment]
Unfortunately this is basically impossible to bisect due to the push_notebook regression in 0.12.11, which is also roughly the same time when this regression occurred.

**Comment by mattpap** [comment]
Related issues: #7497

**Comment by philippjfr** [comment]
Thanks @mattpap, I think it's the same issue really so I'd be happy to close this as a duplicate but I will defer to your judgement.

**Comment by axil** [comment]
> In earlier versions of bokeh (specifically 0.12.10) it was possible to add new bokeh renderers to a plot and then use push_notebook to update the plot.
Even in 0.12.10 there's problem with updating the newly added plots sources though `push_notebook()`:

![image](https://user-images.githubusercontent.com/170910/42568307-dd225ff0-8536-11e8-8a2e-0016219b1fcf.png)

Cell 6 doesn't work.

**Comment by axil** [comment]
ModelChangedEvent is not fired

**Comment by axil** [comment]
Incidentally, `h.doc != line2.data_source.document`. 
Also `h.doc._hold == True` while `line2.data_source.document._hold == None`
Is it on purpose?

**Comment by axil** [comment]
Here's a quick workaround:

![image](https://user-images.githubusercontent.com/170910/42573075-50057a44-8545-11e8-9b27-963bbb454fde.png)

**Comment by axil** [comment]
The same thing with visibility:

    line2.visible = False; push_notebook(h)
only works after 

    line2._document = h.doc

**Comment by axil** [comment]
Is it a bug or a feature? :)

**Comment by bryevdv** [comment]
I'm not sure we can do anything about this. When Bokeh content is displayed in an output cell, the process to make that happens involves cloning all the shown models into a new document of their own, associated with the handle, and using that document to display. That happens at `show` time, and that is why adding things later has no effect, the new glyphs are not part of the handle's document. 

Things I think we can do:

* document the limitations of `push_notebook` better and advertise best-practices more loudly (i.e. prepare the "whole plot" up front, including glyphs with no data yet that you will add data for later)

* add some kind of explicit API that basically does the `foo.doc = h._doc` operation but in a safe, 
 documented and supported way. I don't think there is any way to make this happen automagically.

**Comment by mattpap** [comment]
The original issue is fixed in PR #8085 (by fixing issue #7497; this one is almost a duplicate, though it's good to test that the notebook works as well). The other problem is unrelated and should be moved to a new issue.

**Comment by bryevdv** [comment]
The original issue is definitely fixed

<img width="556" alt="screen shot 2018-09-13 at 09 27 02" src="https://user-images.githubusercontent.com/1078448/45501844-3a0fd300-b737-11e8-976d-ef3ddaf2ca66.png">

But apparently (rto my surprise) the other issue is now working as well:

<img width="492" alt="screen shot 2018-09-13 at 09 28 31" src="https://user-images.githubusercontent.com/1078448/45501923-780cf700-b737-11e8-90ea-8af3394bb898.png">

I'm not sure offhand how that is, or why, but  as everything here is fixed, I will close.

**Comment by bryevdv** [comment]
OK, let's make a separate issue for that narrow issue then. IMO the most important problem here was that adding a glyph after `push_notebook` used to not work, but now it does (I just tried again to make sure)

**Comment by mattpap** [comment]
I curios what fixed it then, because I don't recall doing anything about this issue before PR #8085.

**Comment by bryevdv** [comment]
confirming still fixed with #8085 

![screen shot 2019-01-30 at 19 08 30](https://user-images.githubusercontent.com/1078448/52028068-75caef80-24c2-11e9-8795-2793f9a23906.png)
